kona/src/main.rs

```diff
@@ -0,0 +1,283 @@
+
+#[derive(Debug)]
+pub(crate) struct KonaExEx<Node: FullNodeComponents> {
+    /// The rollup configuration
+    cfg: Arc<RollupConfig>,
+    /// The context of the Execution Extension
+    ctx: ExExContext<Node>,
+    /// The chain provider to follow the chain state based on ExEx events
+    chain_provider: LocalChainProvider,
+    /// The L2 chain provider to fetch L2 block information and optionally
+    /// verify newly derived payloads against
+    l2_provider: AlloyL2ChainProvider,
+    /// The blob provider to fetch blobs from the beacon client
+    blob_provider: ExExBlobProvider,
+    /// The validator to verify newly derived payloads
+    validator: Box<dyn AttributesValidator + Send>,
+    /// The derivation pipeline
+    pipeline: LocalPipeline,
+    /// The current L2 block we are processing
+    l2_block_cursor: L2BlockInfo,
+    /// A map of L1 anchor blocks to their L2 cursor
+    l1_to_l2_block_cursor: HashMap<BlockInfo, L2BlockInfo>,
+    /// Whether the ExEx is synced to the L2 genesis
+    is_synced_to_l2_genesis: bool,
+    /// Whether we should advance the L2 block cursor
+    should_advance_l2_block_cursor: bool,
+    /// The number of derived payloads so far
+    derived_payloads_count: u64,
+}
+
+impl<Node: FullNodeComponents> KonaExEx<Node> {
+    /// Creates a new instance of the Kona Execution Extension.
+    pub async fn new(ctx: ExExContext<Node>, args: KonaArgsExt) -> Self {
+        // TODO: make chain id configurable
+        let cfg = Arc::new(OP_MAINNET_CONFIG);
+        let mut chain_provider = LocalChainProvider::new();
+        chain_provider.insert_l2_genesis_block(cfg.genesis.l1);
+
+        let mut l2_provider = AlloyL2ChainProvider::new_http(args.l2_rpc_url.clone(), cfg.clone());
+        let blob_provider = ExExBlobProvider::new_from_beacon_client(args.beacon_client_url);
+        let dap = EthereumDataSource::new(chain_provider.clone(), blob_provider.clone(), &cfg);
+
+        let attributes =
+            LocalAttributesBuilder::new(cfg.clone(), l2_provider.clone(), chain_provider.clone());
+
+        let validator: Box<dyn AttributesValidator + Send> = match args.validation_mode {
+            ValidationMode::Trusted => Box::new(TrustedValidator::new_http(
+                args.l2_rpc_url,
+                cfg.canyon_time.unwrap_or_default(),
+            )),
+            ValidationMode::EngineApi => Box::new(EngineApiValidator::new_http(
+                // The presence of these variables is enforced by the clap configuration,
+                // so we can safely unwrap them here.
+                args.l2_engine_api_url.unwrap(),
+                args.l2_engine_jwt.unwrap(),
+            )),
+        };
+
+        // TODO: use static genesis block info without fetching it
+        let cursor = l2_provider.l2_block_info_by_number(cfg.genesis.l2.number).await.unwrap();
+        let tip = chain_provider.block_info_by_number(cursor.l1_origin.number).await.unwrap();
```

Comment:
Not sure we can construct this manually without fetching since the `l1_origin` field is a `BlockID` which is a subset of `BlockInfo`, and doesn't contain all the necessary info.

I don't think this will work as is because the `LocalChainProvider` won't have this block information. We need to ask reth for this block information.

Reply:
Would it be acceptable to change `l1_origin` to be `BlockInfo`? 
Alternatively, could we keep a map of all these L1 origin block infos separately somewhere?